client/hwctl/src/main.rs

```diff
@@ -17,32 +17,43 @@
  *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
  */
 
-use anyhow::Result;
-use std::process::exit;
+use anyhow::{Context, Result};
+use std::process;
 
 use hwlib::{
     models::request_validators::{CertificationStatusRequest, Paths},
     send_certification_status_request,
 };
 
 #[tokio::main]
-async fn main() -> Result<()> {
-    let cert_status_request = CertificationStatusRequest::new(Paths::default())?;
-    println!(
-        "Request:\n{}",
-        serde_json::to_string_pretty(&cert_status_request)?
-    );
-    let url = std::env::var("HW_API_URL").unwrap_or_else(|_| String::from("https://hw.ubuntu.com"));
-    let response = send_certification_status_request(url, &cert_status_request).await;
-
-    match response {
-        Ok(response) => {
-            println!("\nResponse:\n{}", serde_json::to_string_pretty(&response)?);
-            exit(0);
-        }
+async fn main() -> process::ExitCode {
+    match run().await {
+        Ok(_) => process::ExitCode::SUCCESS,
         Err(e) => {
-            eprintln!("{}", e);
-            exit(1);
+            eprintln!("ERROR: {e:?}");
+            process::ExitCode::FAILURE
         }
     }
 }
+
+async fn run() -> Result<()> {
+    let cert_status_request = CertificationStatusRequest::new(Paths::default()).context(
+        "cannot collect system data. \
+         Ensure that hwctl is running as root and on an SMBIOS-compatible system.",
+    )?;
+
+    let request_json = serde_json::to_string_pretty(&cert_status_request)
+        .context("cannot serialize request to JSON")?;
+    println!("Request:\n{}", request_json);
+
+    let url = std::env::var("HW_API_URL").unwrap_or_else(|_| String::from("https://hw.ubuntu.com"));
```

Comment:
This isn't a massive issue, but that `String::from` requiring us to use a closure to prevent unnecessary allocations seems a little clunky. Instead, is it possible to handle an `Option<&str>` rather than an `Option<String>`? (I don't recall the type of the first parameter to `send_certification_status_request`)

```rust
let url = env::var("HW_API_URL").as_deref().unwrap_or("https://hw.ubuntu.com");
```
Also, I think it would be slightly nicer to use the `var` function as `env::var` rather than `std::env::var` (the `std` part doesn't seem too important)

Reply:
If I use `&str` instead of `String`, I get this error
`creates a temporary value which is freed while still in use`

Reply:
Ahhh yes of course, the `Option<String> ` from `env::var` would require a binding, my mistake. That seems like too much, no worries, I think this code is probably cleanest as-is (this is a one-time cost anyway)